File: matrix/rotate_matrix.cpp
/**
 * @file
 * @brief Implementation to rotate a matrix by 90 degrees clockwise
 * @details
 * Rotates an N x N matrix 90 degrees clockwise in-place.
 * Space Complexity: O(1)
 * Time Complexity: O(N^2)
 */

#include <algorithm>
#include <cassert>
#include <iostream>
#include <vector>

/**
 * @brief Rotates matrix 90 degrees clockwise
 * @param matrix Reference to the matrix to be rotated
 */
void rotate(std::vector<std::vector<int>> &matrix) {
  int n = matrix.size();

  // Transpose the matrix
  for (int i = 0; i < n; i++) {
    for (int j = i; j < n; j++) {
      std::swap(matrix[i][j], matrix[j][i]);
    }
  }

  // Reverse each row
  for (int i = 0; i < n; i++) {
    std::reverse(matrix[i].begin(), matrix[i].end());
  }
}

/**
 * @brief Self-test implementations
 */
void test() {
  std::vector<std::vector<int>> matrix = {{1, 2, 3}, {4, 5, 6}, {7, 8, 9}};
  std::vector<std::vector<int>> expected = {{7, 4, 1}, {8, 5, 2}, {9, 6, 3}};

  rotate(matrix);
  assert(matrix == expected);

  std::cout << "All tests passed!" << std::endl;
}

int main() {
  test();
  return 0;
}

File: matrix/rotate_matrix.rs
/**
 * Rotate Matrix 90 degrees clockwise
 */

pub fn rotate(matrix: &mut Vec<Vec<i32>>) {
    let n = matrix.len();

    // Transpose
    for i in 0..n {
        for j in i..n {
            let temp = matrix[i][j];
            matrix[i][j] = matrix[j][i];
            matrix[j][i] = temp;
        }
    }

    // Reverse rows
    for i in 0..n {
        matrix[i].reverse();
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_rotate() {
        let mut matrix = vec![
            vec![1, 2, 3],
            vec![4, 5, 6],
            vec![7, 8, 9],
        ];
        let expected = vec![
            vec![7, 4, 1],
            vec![8, 5, 2],
            vec![9, 6, 3],
        ];
        rotate(&mut matrix);
        assert_eq!(matrix, expected);
    }
}

fn main() {
    let mut matrix = vec![
        vec![1, 2, 3],
        vec![4, 5, 6],
        vec![7, 8, 9],
    ];
    rotate(&mut matrix);
    println!("{:?}", matrix);
}
